### PR TITLE
cancancan, groups and members

### PR DESCRIPTION
**Pull request by hamzashariq7**

### CODE REVIEW

File: app/assets/javascripts/groups.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (AyishmAziz) [inline app/assets/javascripts/groups.coffee]: Remove this file , if not using

File: app/assets/javascripts/members.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (AyishmAziz) [inline app/assets/javascripts/members.coffee]: Remove file , if not using

File: app/assets/stylesheets/groups.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Groups controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (AyishmAziz) [inline app/assets/stylesheets/groups.scss]: Remove file , if not using

File: app/assets/stylesheets/members.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Members controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (AyishmAziz) [inline app/assets/stylesheets/members.scss]: Remove file , if not using

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -2,4 +2,30 @@ class ApplicationController < ActionController::Base
...
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+
+  rescue_from CanCan::AccessDenied do |exception|
+    respond_to do |format|
+      format.json { head :forbidden, content_type: 'text/html' }
+      format.html { redirect_to main_app.root_url, notice: exception.message }
+      format.js   { head :forbidden, content_type: 'text/html' }
+    end
+  end
```
Comment (AyishmAziz) [inline app/controllers/application_controller.rb:12]: Are you handling all these formats ?

File: app/controllers/companies_controller.rb [hunk 6]
```diff
@@ -1,2 +1,15 @@
...
+  # end
+  # before_action :authenticate_admin!, only: [:employees, :groups]
+
+  def dashboard
+
+  end
+
+  def company_settings
+
+  end
```
Comment (AyishmAziz) [inline app/controllers/companies_controller.rb:12]: put empty method definitions on a single line

File: app/helpers/groups_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module GroupsHelper
+end
```
Comment (AyishmAziz) [inline app/helpers/groups_helper.rb]: Remove file , if not needed

File: app/helpers/members_helper.rb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+module MembersHelper
+end
```
Comment (AyishmAziz) [inline app/helpers/members_helper.rb]: Remove file , if not needed

File: app/models/group.rb [hunk 9]
```diff
@@ -4,4 +4,8 @@ class Group < ActiveRecord::Base
 
   validates :name, presence: true, uniqueness: true, length: { maximum: 50 }
```
Comment (AyishmAziz) [inline app/models/group.rb]: discuss name length validation with zohaib

File: app/models/user.rb [hunk 10]
```diff
@@ -10,8 +10,20 @@ class User < ActiveRecord::Base
   has_many :surveys
   has_and_belongs_to_many :groups
 
-  validates :first_name, presence: true, uniqueness: true, length: { maximum: 20 }
-  validates :last_name, presence: true, uniqueness: true, length: { maximum: 20 }
+  validates :first_name, presence: true, length: { maximum: 20 }
+  validates :last_name, presence: true, length: { maximum: 20 }
```
Comment (AyishmAziz) [inline app/models/user.rb]: Discuss first_name , last_name length validation with zohaib

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -2,6 +2,18 @@
...
+
+  <!-- jQuery library -->
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
+
+  <!-- Popper JS -->
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
+
+  <!-- Latest compiled JavaScript -->
+  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
+
```
Comment (AyishmAziz) [inline app/views/layouts/application.html.erb]: Indentation

File: db/migrate/20190716162420_create_questions.rb [hunk 12]
```diff
@@ -5,7 +5,8 @@ def change
       t.string :type
       t.integer :survey_id
       t.belongs_to :survey, index: true
-
+      t.belongs_to :company, index: true
+
```
Comment (AyishmAziz) [inline db/migrate/20190716162420_create_questions.rb]: Remove associations from migrations

File: db/migrate/20190716171219_create_surveys.rb [hunk 13]
```diff
@@ -9,6 +9,7 @@ def change
       t.datetime :expiry
       t.integer :user_id
       t.belongs_to :user, index: true
+      t.belongs_to :company, index: true
```
Comment (AyishmAziz) [inline db/migrate/20190716171219_create_surveys.rb]: Remove Associations . Add index

File: spec/controllers/groups_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe GroupsController, type: :controller do
+
+end
```
Comment (AyishmAziz) [inline spec/controllers/groups_controller_spec.rb:34]: Remove file , if not needed

File: spec/views/companies/dashboard.html.erb_spec.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "companies/dashboard.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (AyishmAziz) [inline spec/views/companies/dashboard.html.erb_spec.rb:5]: Remove file , if not needed

File: spec/views/members/create.html.erb_spec.rb [hunk 16]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "members/create.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (AyishmAziz) [inline spec/views/members/create.html.erb_spec.rb:5]: Remove file , if not needed

File: spec/views/members/edit.html.erb_spec.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "members/edit.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (AyishmAziz) [inline spec/views/members/edit.html.erb_spec.rb:5]: Remove file , if not needed

File: spec/views/members/new.html.erb_spec.rb [hunk 18]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "members/new.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (AyishmAziz) [inline spec/views/members/new.html.erb_spec.rb:5]: Remove file , if not needed

File: spec/views/members/update.html.erb_spec.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "members/update.html.erb", type: :view do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (AyishmAziz) [inline spec/views/members/update.html.erb_spec.rb:5]: Remove file , if not needed

File: db/migrate/20190716164758_create_responses.rb [hunk 20]
```diff
@@ -4,6 +4,7 @@ def change
       t.string :answer
       t.integer :question_id
       t.belongs_to :question, index: true
+      t.belongs_to :company, index: true
 
       t.timestamps null: false
     end
```
Comment (AyishmAziz) [inline db/migrate/20190716164758_create_responses.rb:9]: Remove associations

File: app/controllers/application_controller.rb [hunk 21]
```diff
@@ -2,4 +2,30 @@ class ApplicationController < ActionController::Base
...
+  before_action :authenticate_user! do
+    @current_company ||= current_user.company
+  end
+
+  def after_sign_in_path_for(resource)
+    dashboard_company_path(current_user.company)
+  end
+
+  def company_employees
+    @company.users
```
Comment (zohaibbutt13) [inline app/controllers/application_controller.rb]: do we need this method ?

File: app/controllers/groups_controller.rb [hunk 22]
```diff
@@ -0,0 +1,43 @@
+class GroupsController < ApplicationController
+  load_and_authorize_resource
+
+  before_action do
+    @employees = User.accessible_by(current_ability)
```
Comment (zohaibbutt13) [inline app/controllers/groups_controller.rb:5]: why do we need employees before every action ?

File: app/controllers/groups_controller.rb [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def new
+  end
+
+  def create
+    @group.company = @current_company
```
Comment (zohaibbutt13) [inline app/controllers/groups_controller.rb]: we only need company_id during save and write a method in model to save object (model responsibility)
and handle errors in models and then return to controller then set messages in flash

File: app/controllers/users/registrations_controller.rb [hunk 24]
```diff
@@ -10,9 +12,28 @@ class Users::RegistrationsController < Devise::RegistrationsController
   # end
 
   # POST /resource
-  # def create
-  #   super
-  # end
+  def create
+    build_resource(sign_up_params)
+    resource.company = Company.find(3)
```
Comment (zohaibbutt13) [inline app/controllers/users/registrations_controller.rb]: change with @current_company

File: app/models/ability.rb [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+    
+    if user.admin?
+      can :manage, [User, Group], company_id: user.company_id
+      cannot [:destroy, :edit], user
```
Comment (zohaibbutt13) [inline app/models/ability.rb:11]: why admin cannot destroy and edit user ?

File: app/models/ability.rb [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    elsif user.supervisor
```
Comment (zohaibbutt13) [inline app/models/ability.rb]: question mark

File: app/models/ability.rb [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+      can :read, [User, Group], company_id: user.company_id
+    elsif user.member
+      can :read, [User, Group], company_id: user.company_id
```
Comment (zohaibbutt13) [inline app/models/ability.rb:13]: member should not read users and groups 
member can only fill surveys which they are allowed to

File: app/models/ability.rb [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+    else
+
+    end
+    # Define abilities for the passed in user here. For example:
```
Comment (zohaibbutt13) [inline app/models/ability.rb]: remove all this commented code

File: app/models/group.rb [hunk 29]
```diff
@@ -2,6 +2,10 @@ class Group < ActiveRecord::Base
   belongs_to :company
   has_and_belongs_to_many :users
 
-  validates :name, presence: true, uniqueness: true, length: { maximum: 50 }
+  validates :name, presence: true, uniqueness: true, length: { maximum: 150 }
   validates :description, presence: true, length: { maximum: 500 }
+  validates :users, length: {
+    minimum: 1,
+    message: 'A group should have at least 1 members'
```
Comment (zohaibbutt13) [inline app/models/group.rb]: members -> member
group -> Group

File: app/models/user.rb [hunk 30]
```diff
@@ -5,13 +5,31 @@ class User < ActiveRecord::Base
...
+
+  def member?
+    role == User::ROLE[:member]
+  end
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
 
   ROLE = {admin: 'admin', supervisor: 'supervisor', member: 'member'}
```
Comment (zohaibbutt13) [inline app/models/user.rb]: define constants at the start of file

File: app/views/companies/dashboard.html.erb [hunk 31]
```diff
@@ -0,0 +1,7 @@
+<h1>Companies#dashboard</h1>
+<p>Find me in app/views/companies/dashboard.html.erb</p>
+
+<ul>
+  <%= link_to "<li>Employees</li>".html_safe, members_path %>
```
Comment (zohaibbutt13) [inline app/views/companies/dashboard.html.erb]: use block for link_to

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 32]
```diff
@@ -1,4 +1,8 @@
-<p>Welcome <%= @email %>!</p>
+<p>Welcome <%= @resource.full_name %>!</p>
+
+<% unless @resource.admin? %>
+<p> Your password is <%= @resource.initial_password %>.</p>
```
Comment (zohaibbutt13) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: indentation

File: app/views/groups/_fields.html.erb [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+
+<%= form_for locals[:group] do |f| %>
+  <div class="field">
+    <%= f.label :name %><br />
+    <%= f.text_field :name, autofocus: true, autocomplete: "name" %>
+  </div>
+
+  <div class="form-group">
+    <%= f.label :description %><br />
+    <%= f.text_area :description, autofocus: true, class: 'form-control', autocomplete: "description" %>
```
Comment (zohaibbutt13) [inline app/views/groups/_fields.html.erb:11]: specify number of rows

File: app/views/members/_form.html.erb [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+<div class="field">
+  <%= f.label :role %><br />
+  <%= f.select :role, User::ROLE.reject { |key, value| key == :admin } %>
+  <br>
+</div>
+
+<div class="actions">
+  <%= f.submit locals[:submit] %>
+</div>
+
```
Comment (zohaibbutt13) [inline app/views/members/_form.html.erb]: remove extra spaces and lines

File: app/views/members/edit.html.erb [hunk 35]
```diff
@@ -0,0 +1,5 @@
+<h2>Edit Member</h2>
+
+<%= form_for @member, url: member_path(@member), method: :put do |f| %>
```
Comment (zohaibbutt13) [inline app/views/members/edit.html.erb]: you can use form_for in _form.html.erb for both cases new and edit 
method should be put or patch?

File: app/views/members/index.html.erb [hunk 36]
```diff
@@ -0,0 +1,23 @@
+<h1> Members </h1>
+
+<% if can? :manage, User %>
+  <td><%= link_to "Add new", new_member_path %></td>
```
Comment (zohaibbutt13) [inline app/views/members/index.html.erb]: add new what?

Comment (zohaibbutt13) [inline app/views/members/index.html.erb]: add new -> Create new User

File: app/views/members/index.html.erb [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+<% end %>
+
+<table>
+  <% @members.each do |member| %>
+    <tr>
+      <td><%= "#{member.first_name} #{member.last_name}" %></td>
```
Comment (zohaibbutt13) [inline app/views/members/index.html.erb]: {} should have spaces at start and end
you can use full_name method

File: app/views/members/index.html.erb [hunk 38]
```diff
@@ -0,0 +1,23 @@
...
+      <% if can? :manage, member %>
```
Comment (zohaibbutt13) [inline app/views/members/index.html.erb:11]: no need for this line when you are explicitly checking the ability of each action

File: app/views/members/index.html.erb [hunk 39]
```diff
@@ -0,0 +1,23 @@
...
+        <% unless cannot? :edit, member %>
```
Comment (zohaibbutt13) [inline app/views/members/index.html.erb:12]: use if can?

File: app/views/members/index.html.erb [hunk 40]
```diff
@@ -0,0 +1,23 @@
...
+          <td><%= link_to 'Edit', edit_member_path(member) %></td>
+        <% end %>
+        <% unless cannot? :destroy, member %>
+          <td><%= link_to 'Delete', member_path(member), method: :delete, data: { confirm: 'You sure?' } %>
```
Comment (zohaibbutt13) [inline app/views/members/index.html.erb]: review confirm text with other teams members as well

File: db/migrate/20190722123136_add_initial_password_to_users.rb [hunk 41]
```diff
@@ -0,0 +1,5 @@
+class AddInitialPasswordToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :initial_password, :string
```
Comment (zohaibbutt13) [inline db/migrate/20190722123136_add_initial_password_to_users.rb:3]: no need of this new column 
in your case when you create new user then update its password with random generated value

File: spec/helpers/groups_helper_spec.rb [hunk 42]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (zohaibbutt13) [inline spec/helpers/groups_helper_spec.rb:1]: remove this file if you are going to write test cases for views

File: spec/views/companies/dashboard.html.erb_spec.rb [hunk 43]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (zohaibbutt13) [inline spec/views/companies/dashboard.html.erb_spec.rb:1]: remove this file

File: spec/views/members/create.html.erb_spec.rb [hunk 44]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (zohaibbutt13) [inline spec/views/members/create.html.erb_spec.rb:1]: remove this files

File: survey.sh [hunk 45]
```diff
@@ -0,0 +1 @@
+cp config/database.yml.survey config/database.yml
```
Comment (zohaibbutt13) [inline survey.sh]: remove this files no need it anymore

**Pull request merged**